src/python/pants/backend/helm/dependency_inference/deployment.py

```diff
@@ -0,0 +1,180 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+from pathlib import PurePath
+
+from pants.backend.docker.target_types import AllDockerImageTargets
+from pants.backend.docker.target_types import rules as docker_target_types_rules
+from pants.backend.helm.subsystems import k8s_parser
+from pants.backend.helm.subsystems.k8s_parser import ParsedKubeManifest, ParseKubeManifestRequest
+from pants.backend.helm.target_types import (
+    AllHelmDeploymentTargets,
+    HelmDeploymentDependenciesField,
+    HelmDeploymentFieldSet,
+)
+from pants.backend.helm.target_types import rules as helm_target_types_rules
+from pants.backend.helm.util_rules import renderer
+from pants.backend.helm.util_rules.renderer import (
+    HelmDeploymentCmd,
+    HelmDeploymentRequest,
+    RenderedHelmFiles,
+)
+from pants.backend.helm.utils.docker import ImageRef
+from pants.backend.helm.utils.yaml import FrozenYamlIndex, MutableYamlIndex
+from pants.engine.addresses import Address
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    DependenciesRequest,
+    ExplicitlyProvidedDependencies,
+    FieldSet,
+    InferDependenciesRequest,
+    InferredDependencies,
+)
+from pants.engine.unions import UnionRule
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+from pants.util.ordered_set import FrozenOrderedSet, OrderedSet
+from pants.util.strutil import pluralize, softwrap
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class HelmDeploymentReport:
+    address: Address
+    image_refs: FrozenYamlIndex[ImageRef]
+
+    @property
+    def all_image_refs(self) -> FrozenOrderedSet[ImageRef]:
+        return FrozenOrderedSet(self.image_refs.values())
+
+
+@rule(desc="Analyse Helm deployment", level=LogLevel.DEBUG)
+async def analyse_deployment(field_set: HelmDeploymentFieldSet) -> HelmDeploymentReport:
+    rendered_deployment = await Get(
+        RenderedHelmFiles,
+        HelmDeploymentRequest(
+            cmd=HelmDeploymentCmd.RENDER,
+            field_set=field_set,
+            description=f"Rendering Helm deployment {field_set.address}",
+        ),
+    )
+
+    parsed_manifests = await MultiGet(
+        Get(
+            ParsedKubeManifest,
+            ParseKubeManifestRequest(filename=file, digest=rendered_deployment.snapshot.digest),
```

Comment:
Because the whole snapshot is going in, the cache key will still end up depending on all of the files in the snapshot (and the extraction will be able to see all files). Is that expected?

Reply:
more like a consequence than actually intended.

I initially wrote this using breaking it into smaller `Digest` using `DigestSubset` for each of the files and passing that to the process that parses it. However it did look like a bit overkill. Maybe I could just ask for the `DigestContents` and transform each `FileDigest` into a `Digest` for the underlying process. That should have a better usage of cache, right?